Failed to compile on Mac OS Mojave 10.14.5 (18F132)

I tried installing accupy and compilation failed:
```  
~/partitioning: python3 -m pip install accupy
Collecting accupy
...
    Complete output from command /usr/local/bin/python3 -u -c "import setuptools, tokenize;__file__='/private/var/folders/_j/_5vgryt961bg8tw_2jvgmc6w0000gp/T/pip-install-t2i_vxvy/pyfma/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace([87/1791]
\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/var/folders/_j/_5vgryt961bg8tw_2jvgmc6w0000gp/T/pip-record-38rgrthc/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build/lib.macosx-10.9-x86_64-3.7
    creating build/lib.macosx-10.9-x86_64-3.7/pyfma
    copying pyfma/__init__.py -> build/lib.macosx-10.9-x86_64-3.7/pyfma
    copying pyfma/__about__.py -> build/lib.macosx-10.9-x86_64-3.7/pyfma
    running build_ext
    building '_pyfma' extension
    creating build/temp.macosx-10.9-x86_64-3.7
    creating build/temp.macosx-10.9-x86_64-3.7/src
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -arch x86_64 -g -I/Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m -I/Users/smukherjee/Library/Python/3.7/include/python3.7m -I/Library/Fr
ameworks/Python.framework/Versions/3.7/include/python3.7m -c src/pybind11.cpp -o build/temp.macosx-10.9-x86_64-3.7/src/pybind11.o
    In file included from src/pybind11.cpp:3:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/pybind11.h:44:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/attr.h:13:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/cast.h:13:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/pytypes.h:12:
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:298:17: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    using ssize_t = Py_ssize_t;
                    ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:299:17: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    using size_t  = std::size_t;
                    ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:302:6: warning: scoped enumerations are a C++11 extension [-Wc++11-extensions]
    enum class return_value_policy : uint8_t {
         ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:302:12: error: ISO C++ forbids forward references to 'enum' types
    enum class return_value_policy : uint8_t {
               ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:302:32: error: expected unqualified-id
    enum class return_value_policy : uint8_t {
                                   ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:355:15: error: unknown type name 'constexpr'
    inline static constexpr int log2(size_t n, int k = 0) { return (n <= 1) ? k : log2(n >> 1, k + 1); }
                  ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:355:25: error: expected unqualified-id
    inline static constexpr int log2(size_t n, int k = 0) { return (n <= 1) ? k : log2(n >> 1, k + 1); }
                            ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:358:15: error: unknown type name 'constexpr'
    inline static constexpr size_t size_in_ptrs(size_t s) { return 1 + ((s - 1) >> log2(sizeof(void *))); }
                  ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:358:1: warning: inline variables are a C++17 extension [-Wc++17-extensions]
    inline static constexpr size_t size_in_ptrs(size_t s) { return 1 + ((s - 1) >> log2(sizeof(void *))); }
    ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:358:31: error: expected ';' after top level declarator
    inline static constexpr size_t size_in_ptrs(size_t s) { return 1 + ((s - 1) >> log2(sizeof(void *))); }
                                  ^
                                  ;
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:386:39: error: use of undeclared identifier 'instance_simple_holder_in_ptrs'
            void *simple_value_holder[1 + instance_simple_holder_in_ptrs()];
                                          ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:433:49: error: unknown type name 'type_info'; did you mean 'std::type_info'?
        value_and_holder get_value_and_holder(const type_info *find_type = nullptr, bool throw_if_missing = true);
                                                    ^~~~~~~~~
                                                    std::type_info
    /Library/Developer/CommandLineTools/usr/include/c++/v1/typeinfo:87:29: note: 'std::type_info' declared here                                                                                                                                                      [28/1791]
    class _LIBCPP_EXCEPTION_ABI type_info
                                ^
    In file included from src/pybind11.cpp:3:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/pybind11.h:44:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/attr.h:13:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/cast.h:13:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/pytypes.h:12:
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:436:12: error: unknown type name 'constexpr'
        static constexpr uint8_t status_holder_constructed  = 1;
               ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:436:29: error: expected ';' at end of declaration list
        static constexpr uint8_t status_holder_constructed  = 1;
                                ^
                                ;
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:437:12: error: unknown type name 'constexpr'
        static constexpr uint8_t status_instance_registered = 2;
               ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:437:29: error: expected ';' at end of declaration list
        static constexpr uint8_t status_instance_registered = 2;
                                ^
                                ;
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:449:58: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    template <bool B, typename T = void> using enable_if_t = typename std::enable_if<B, T>::type;
                                                             ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:450:65: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    template <bool B, typename T, typename F> using conditional_t = typename std::conditional<B, T, F>::type;
                                                                    ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:451:43: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    template <typename T> using remove_cv_t = typename std::remove_cv<T>::type;
                                              ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:452:50: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    template <typename T> using remove_reference_t = typename std::remove_reference<T>::type;
                                                     ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:460:10: error: unknown type name 'size_t'
    template<size_t ...> struct index_sequence  { };
             ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:460:17: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
    template<size_t ...> struct index_sequence  { };
                    ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:461:10: error: unknown type name 'size_t'
    template<size_t N, size_t ...S> struct make_index_sequence_impl : make_index_sequence_impl <N - 1, N - 1, S...> { };
             ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:461:20: error: unknown type name 'size_t'
    template<size_t N, size_t ...S> struct make_index_sequence_impl : make_index_sequence_impl <N - 1, N - 1, S...> { };
                       ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:461:27: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
    template<size_t N, size_t ...S> struct make_index_sequence_impl : make_index_sequence_impl <N - 1, N - 1, S...> { };
                              ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:462:10: error: unknown type name 'size_t'
    template<size_t ...S> struct make_index_sequence_impl <0, S...> { typedef index_sequence<S...> type; };
             ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:462:17: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
    template<size_t ...S> struct make_index_sequence_impl <0, S...> { typedef index_sequence<S...> type; };
                    ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:463:10: error: unknown type name 'size_t'
    template<size_t N> using make_index_sequence = typename make_index_sequence_impl<N>::type;
             ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:463:48: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    template<size_t N> using make_index_sequence = typename make_index_sequence_impl<N>::type;
                                                   ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:467:32: error: expected template parameter
    template <typename ISeq, size_t, bool...> struct select_indices_impl { using type = ISeq; };
                                   ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:467:38: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
    template <typename ISeq, size_t, bool...> struct select_indices_impl { using type = ISeq; };
                                         ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:467:85: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    template <typename ISeq, size_t, bool...> struct select_indices_impl { using type = ISeq; };
                                                                                        ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:468:11: error: unknown type name 'size_t'
    template <size_t... IPrev, size_t I, bool B, bool... Bs> struct select_indices_impl<index_sequence<IPrev...>, I, B, Bs...>
              ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:468:17: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
    template <size_t... IPrev, size_t I, bool B, bool... Bs> struct select_indices_impl<index_sequence<IPrev...>, I, B, Bs...>
                    ^
    fatal error: too many errors emitted, stopping now [-ferror-limit=]
    15 warnings and 20 errors generated.
    error: command 'gcc' failed with exit status 1
```

It looks like one of the problems is that my compiler is not expecting C++11. I'm going to try to see if if I can pass a flag to gcc to get it to accept C++ 11.

--- a/setup.py
+++ b/setup.py
@@ -1,14 +1,16 @@
+import sys
+
+import setuptools
 from setuptools import Extension, find_packages, setup
+from setuptools.command.build_ext import build_ext
 
 
+# https://github.com/pybind/python_example/
 class get_pybind_include:
-    def __init__(self, user=False):
-        self.user = user
-
     def __str__(self):
         import pybind11
 
-        return pybind11.get_include(self.user)
+        return pybind11.get_include()
 
 
 ext_modules = [
@@ -16,13 +18,75 @@ ext_modules = [
         "_accupy",
         ["src/pybind11.cpp"],
         language="c++",
-        include_dirs=[
-            "/usr/include/eigen3/",
-            get_pybind_include(),
-            get_pybind_include(user=True),
-        ],
+        include_dirs=["/usr/include/eigen3/", get_pybind_include()],
     )
 ]
+
+
+# cf http://bugs.python.org/issue26689
+def has_flag(compiler, flagname):
+    """Return a boolean indicating whether a flag name is supported on
+    the specified compiler.
+    """
+    import tempfile
+    import os
+
+    with tempfile.NamedTemporaryFile("w", suffix=".cpp", delete=False) as f:
+        f.write("int main (int argc, char **argv) { return 0; }")
+        fname = f.name
+    try:
+        compiler.compile([fname], extra_postargs=[flagname])
+    except setuptools.distutils.errors.CompileError:
+        return False
+    finally:
+        try:
+            os.remove(fname)
+        except OSError:
+            pass
+    return True
+
+
+def cpp_flag(compiler):
+    flags = ["-std=c++17", "-std=c++14", "-std=c++11"]
+
+    for flag in flags:
+        if has_flag(compiler, flag):
+            return flag
+
+    raise RuntimeError("Unsupported compiler -- at least C++11 support is needed!")
+
+
+class BuildExt(build_ext):
+    c_opts = {
+        "msvc": ["/EHsc"],
+        "unix": [],
+    }
+    l_opts = {
+        "msvc": [],
+        "unix": [],
+    }
+
+    if sys.platform == "darwin":
+        darwin_opts = ["-stdlib=libc++", "-mmacosx-version-min=10.7"]
+        c_opts["unix"] += darwin_opts
+        l_opts["unix"] += darwin_opts
+
+    def build_extensions(self):
+        ct = self.compiler.compiler_type
+        opts = self.c_opts.get(ct, [])
+        link_opts = self.l_opts.get(ct, [])
+        if ct == "unix":
+            opts.append(cpp_flag(self.compiler))
+            if has_flag(self.compiler, "-fvisibility=hidden"):
+                opts.append("-fvisibility=hidden")
+
+        for ext in self.extensions:
+            ext.define_macros = [
+                ("VERSION_INFO", '"{}"'.format(self.distribution.get_version()))
+            ]
+            ext.extra_compile_args = opts
+            ext.extra_link_args = link_opts
+        build_ext.build_extensions(self)
 
 
 setup(
@@ -38,9 +102,11 @@ setup(
         "importlib_metadata",
         "mpmath",
         "numpy",
-        "pybind11 >= 2.2",
+        "pybind11 >= 2.5.0",
         "pyfma",
     ],
+    setup_requires=["pybind11 >= 2.5.0"],
+    cmdclass={"build_ext": BuildExt},
     python_requires=">=3.6",
     description="Accurate sums and dot products for Python",
     long_description=open("README.md").read(),

--- a/setup.py
+++ b/setup.py
@@ -91,7 +91,7 @@ class BuildExt(build_ext):
 
 setup(
     name="accupy",
-    version="0.3.0",
+    version="0.3.1",
     packages=find_packages(),
     ext_modules=ext_modules,
     url="https://github.com/nschloe/accupy",

--- a/test/test_dot.py
+++ b/test/test_dot.py
@@ -92,7 +92,6 @@ def test_speed_comparison1(n_range=None):
         xlabel="n",
         logx=True,
         logy=True,
-        automatic_order=False,
     )
     plt.gca().set_aspect(0.2)
     lgd = plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.0)
@@ -126,7 +125,6 @@ def test_speed_comparison2(n_range=None):
         xlabel="n",
         logx=True,
         logy=True,
-        automatic_order=False,
     )
     plt.gca().set_aspect(0.2)
     lgd = plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.0)

--- a/test/test_sums.py
+++ b/test/test_sums.py
@@ -112,7 +112,6 @@ def test_speed_comparison1(n_range=None):
         xlabel="n",
         logx=True,
         logy=True,
-        automatic_order=False,
     )
     plt.gca().set_aspect(0.5)
     lgd = plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.0)
@@ -155,7 +154,6 @@ def test_speed_comparison2(n_range=None):
         xlabel="n",
         logx=True,
         logy=True,
-        automatic_order=False,
     )
     plt.gca().set_aspect(0.5)
     lgd = plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.0)

